optee/core.c

```diff
@@ -723,20 +782,71 @@ static int optee_probe(struct platform_device *pdev)
 }
 
 static const struct of_device_id optee_dt_match[] = {
-	{ .compatible = "linaro,optee-tz" },
+	{ .compatible = "optee-tz" },
 	{},
 };
 MODULE_DEVICE_TABLE(of, optee_dt_match);
 
+static struct device_node optee_dev_node = {
+	.name = "optee-tz",
+};
+
+void optee_dev_release(struct device *dev)
+{
+	return;
+}
+
+static struct platform_device optee_platform_dev = {
+		.name = "optee-tz",
+		.id = -1,
+		.num_resources = 0,
+		.dev = {
+			.release = optee_dev_release,
+			.of_node = &optee_dev_node,
+	},
+};
+
+static struct platform_device *optee_devices[] __initdata = {
+	&optee_platform_dev
+};
+
 static struct platform_driver optee_driver = {
 	.probe  = optee_probe,
 	.remove = optee_remove,
 	.driver = {
-		.name = "optee",
+		.name = "optee-tz",
 		.of_match_table = optee_dt_match,
 	},
 };
-module_platform_driver(optee_driver);
+
+static int __init optee_drv_init(void)
+{
+	int ret;
+
+#ifndef CONFIG_OF_EARLY_FLATTREE
+	ret = platform_add_devices(optee_devices, ARRAY_SIZE(optee_devices));
+	if (ret) {
+		printk(KERN_ERR "%s(): platform_add_devices() failed, ret %d\n", __func__, ret);
+		return ret;
+	}
+#endif
+
+	ret = platform_driver_register(&optee_driver);
+
+	return ret;
+}
+
+static void __exit optee_drv_exit(void)
+{
+	platform_driver_unregister(&optee_driver);
+
+#ifndef CONFIG_OF_EARLY_FLATTREE
+		platform_device_unregister(&optee_platform_dev);
+#endif
+}
+
+module_init(optee_drv_init);
```

Comment:
We still can't use module_platform_driver(optee_driver)?

Reply:
The driver has been registered/unregistered in device initialize/remove process.